### ISSUE TITLE
Support for CorrelationIdPlugin and CorrelationIdPlugin still executing when application encounters an exception

### ISSUE THREAD

**Issue by derekbekoe** [tomwojcik/starlette-context#9]
**Issue**

When a route raises an exception and so returns a 500 Internal Server Error, x-request-id and x-correlation-id are not set.

The culprit seems to be
https://github.com/tomwojcik/starlette-context/blob/79aa8ffe5b50db2263e2073fc5d252bf442f150c/starlette_context/middleware.py#L46-L52

This is similar to this comment - https://github.com/tiangolo/fastapi/issues/397#issuecomment-543136587

**Expectation**
I'd expect those headers to be set for all responses. One benefit of correlation ids is when there are errors, we can use the ids to track down the issue.

**Steps to reproduce**
Sample.
The "/" endpoint returns the x-request-id and x-correlation-id headers.
The "/error" does not.

```python
from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.requests import Request
from starlette.responses import JSONResponse

import uvicorn
from starlette_context import context, plugins
from starlette_context.middleware import ContextMiddleware

middleware = [
    Middleware(
        ContextMiddleware,
        plugins=(plugins.RequestIdPlugin(), plugins.CorrelationIdPlugin()),
    )
]

app = Starlette(debug=True, middleware=middleware)


@app.route("/")
async def index(request: Request):
    return JSONResponse(context.data)

@app.route("/error")
async def index(request: Request):
    raise RuntimeError()
    return JSONResponse(context.data)

uvicorn.run(app, host="0.0.0.0")
```

**Comment by tomwojcik** [comment]
Frankly, I don't understand. I mean, it can't return anything from the route if the exception has been raised. You either handle that exception. If the exception is not handled then it won't proceed and it won't return the context.

And for 
> When a route raises an exception and so returns a 500 Internal Server Error, x-request-id and x-correlation-id are not set.

Here's a quick proof that it is set.

`from starlette.applications import Starlette
from starlette.middleware import Middleware
from starlette.responses import JSONResponse

import uvicorn
from starlette_context import context, plugins
from starlette_context import _request_scope_context_storage

from starlette_context.middleware import *


class CustomMiddleware(ContextMiddleware):
    def check_if_context_is_set(self):
        import logging
        logger = logging.getLogger()
        from starlette_context import context
        logger.error(f'{context.data}, hi from middleware')

    async def dispatch(
            self, request: Request, call_next: RequestResponseEndpoint
    ) -> Response:
        _starlette_context_token: Token = _request_scope_context_storage.set(
            await self.set_context(request)
        )
        try:
            response = await call_next(request)
            for plugin in self.plugins:
                await plugin.enrich_response(response)

        finally:
            self.check_if_context_is_set()
            _request_scope_context_storage.reset(_starlette_context_token)

        return response


middleware = [
    Middleware(
        CustomMiddleware,
        plugins=(plugins.RequestIdPlugin(), plugins.CorrelationIdPlugin()),
    )
]

app = Starlette(debug=True, middleware=middleware)


@app.route("/")
async def index(request: Request):
    return JSONResponse(context.data)

@app.route("/error")
async def index(request: Request):
    raise RuntimeError()
    return JSONResponse(context.data)


uvicorn.run(app, host="0.0.0.0")`

The existence of context is not influenced by the route as it's created and destroyed in the middleware.

**Comment by tomwojcik** [comment]
I removed my previous answer. Of course, you're right! Thanks for pointing that out.
Let me take care of that myself. I will fix this over the weekend.

Good spot!

**Comment by tomwojcik** [comment]
I removed my answers twice. At the beginning I thought there's nothing to fix, then I've noticed what the issue was, but now I still think there's nothing to fix.

So the exception is raised. What does Starlette do about it?
You either use exception handlers or not. If not, then indeed - it won't return what you want. 

But hey, we all do you them, right? So `ExceptionMiddleware` [docs](https://www.starlette.io/middleware/) allows you to use `exception_handler` per exception. 

> ExceptionMiddleware - Adds exception handlers, so that particular types of expected exception cases can be associated with handler functions. For example raising HTTPException(status_code=404) within an endpoint will end up rendering a custom 404 page.

So basically, you need to add `ContextMiddleware`, then add `ExceptionMiddleware` and in your handler functions context should be available. 

I will write some tests for that to make sure it works like that.

**Comment by derekbekoe** [comment]
Yes the headers x-request-id and x-correlation-id being set in **responses** was what I meant. Updated the initial issue comment to fix the typo.

**Comment by tomwojcik** [comment]
I was digging into it for some time and I don't think there's a clean solution for that. We may discuss things or you might want to open a PR but IMO the best way to do is to figure out in Starlette how to handle it differently. And open a PR or discuss those things there.

I will leave this ticket open.

**Comment by derekbekoe** [comment]
I will close this issue. Thanks for the explanation.

**Comment by tomwojcik** [comment]
@derekbekoe this case is now supported. https://github.com/tomwojcik/starlette-context/pull/79/files
